DynamicModeler/Logic/vtkSlicerDynamicModelerExtrudeTool.cxx

```diff
@@ -189,51 +206,103 @@ bool vtkSlicerDynamicModelerExtrudeTool::RunInternal(vtkMRMLDynamicModelerNode*
+  this->NormalsFilter->Update();
+
   vtkMRMLMarkupsNode* markupsNode = vtkMRMLMarkupsNode::SafeDownCast(surfaceEditorNode->GetNodeReference(EXTRUDE_INPUT_MARKUPS_REFERENCE_ROLE));
 
-  vtkMRMLMarkupsFiducialNode* markupsFiducialNode = vtkMRMLMarkupsFiducialNode::SafeDownCast(markupsNode);
-  vtkMRMLMarkupsLineNode* markupsLineNode = vtkMRMLMarkupsLineNode::SafeDownCast(markupsNode);
-  vtkMRMLMarkupsPlaneNode* markupsPlaneNode = vtkMRMLMarkupsPlaneNode::SafeDownCast(markupsNode);
+  double extrusionValue = this->GetNthInputParameterValue(1, surfaceEditorNode).ToDouble();
 
-  if (markupsFiducialNode)
+  if (markupsNode == nullptr)
   {
-    double center[3] = { 0,0,0 };
-    int controlPointIndex = markupsFiducialNode->GetNthControlPointIndexByPositionStatus(0, vtkMRMLMarkupsNode::PositionDefined);
-    if (markupsFiducialNode >= 0)
-    {
-      markupsFiducialNode->GetNthControlPointPosition(0, center);
-    }
-    this->ExtrudeFilter->SetExtrusionPoint(center);
-    this->ExtrudeFilter->SetExtrusionTypeToPointExtrusion();
+    this->ExtrudeFilter->SetExtrusionTypeToNormalExtrusion();
+    this->ExtrudeFilter->SetScaleFactor(extrusionValue);
   }
-  else if (markupsLineNode || markupsPlaneNode)
+  else
   {
-    double startToEndVector[3] = { 1.0, 0.0, 0.0 };
-    if (markupsLineNode)
+    std::string selectionAlgorithm = this->GetNthInputParameterValue(0, surfaceEditorNode).ToString();
+
+    vtkMRMLMarkupsFiducialNode* markupsFiducialNode = vtkMRMLMarkupsFiducialNode::SafeDownCast(markupsNode);
+    vtkMRMLMarkupsLineNode* markupsLineNode = vtkMRMLMarkupsLineNode::SafeDownCast(markupsNode);
+    vtkMRMLMarkupsPlaneNode* markupsPlaneNode = vtkMRMLMarkupsPlaneNode::SafeDownCast(markupsNode);
+    int numberOfControlPoints = markupsNode->GetNumberOfControlPoints();
+    
+    if (markupsPlaneNode)
     {
-      double startPos[3] = { 0,0,0 };
-      double endPos[3] = { 1,0,0 };
-      markupsLineNode->GetLineStartPositionWorld(startPos);
-      markupsLineNode->GetLineEndPositionWorld(endPos);
-      // startToEndVector = endPos - startPos
-      vtkMath::Subtract(endPos, startPos, startToEndVector);
+      this->ExtrudeFilter->SetScaleFactor(extrusionValue);
+      int planeType = markupsPlaneNode->GetPlaneType();
+      if (
+        ((numberOfControlPoints == 1) && (planeType == vtkMRMLMarkupsPlaneNode::PlaneTypePointNormal)) or
+        ((numberOfControlPoints == 3) && (planeType == vtkMRMLMarkupsPlaneNode::PlaneType3Points)) or
+        ((numberOfControlPoints >= 3) && (planeType == vtkMRMLMarkupsPlaneNode::PlaneTypePlaneFit))
+      )
+      {
+        double startToEndVector[3] = { 1.0, 0.0, 0.0 };
+        markupsPlaneNode->GetNormalWorld(startToEndVector);
+        this->ExtrudeFilter->SetVector(startToEndVector);
+        this->ExtrudeFilter->SetExtrusionTypeToVectorExtrusion();
+      }
     }
-    else if (markupsPlaneNode)
+    
+    if (selectionAlgorithm == "Fixed")
```

Comment:
```suggestion
    if (selectionAlgorithm == "Absolute")
```

Reply:
There may be an issue with internationalization. It seems that the parameter value is used for display as well, so you could not translate. Using a checkbox ("Absolute") instead of combobox would solve this.

Reply:
Ok. I'll switch to a checkbox